feat: show highest permission role when join community

yqrashawn opened the pull request:
fixes #18127
fixes #18581


### Summary

feature behind community-accounts-selection-enabled? flag

- [x] community overview screen
- [x] request to join screen
- [x] address selection screen
- [ ] all permissions drawer -> this drawer is not implemented yet
- [x] get revealed accounts (shared addrs) once joined

rely on 
- #18465
- #18549 (picked into this PR)
- #18571

[status-im/status-mobile/assets/15090582/daedf620-7e16-40be-b9f8-3d238afa7d1d](https://github.com/status-im/status-mobile/assets/15090582/daedf620-7e16-40be-b9f8-3d238afa7d1d)

yqrashawn committed:
feat: show highest permission role in overview and req to join

Signed-off-by: yqrashawn <namy.19@gmail.com>

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | e911490a | [#1](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-18497/1/) | 2024-01-15 08:52:31 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-18497/1/consoleText) |

yqrashawn committed:
feat: get revealed accounts for joined community

Signed-off-by: yqrashawn <namy.19@gmail.com>

ajayesivan committed:
Show token requirements only when membership permissions are present.

- Add membership-permissions? to community state.

yqrashawn committed:
feat: get highest perm role in addr for perms

Signed-off-by: yqrashawn <namy.19@gmail.com>

yqrashawn committed:
feat: ui for address for perm drawer

yqrashawn committed:
test: unit test and spec for handle-community

Signed-off-by: yqrashawn <namy.19@gmail.com>

yqrashawn committed:
fix: for #18581

ilmotta commented on src/status_im/contexts/communities/actions/accounts_selection/view.cljs:
```diff
@@ -4,27 +4,34 @@
+        {:keys [highest-permission-role]} (rf/sub [:community/token-gated-overview id])
+        highest-permission-role-text
+        (i18n/label
+         (condp = highest-permission-role
```
Small refactoring suggestion: extract to a function `role->translation-key`

yqrashawn replied:
added to status-im.contexts.communities.utils

ilmotta commented on src/status_im/contexts/communities/actions/accounts_selection/view.cljs:
```diff
@@ -4,27 +4,34 @@
+        airdrop-account (rf/sub [:communities/airdrop-account id])
+        selected-accounts (rf/sub [:communities/selected-permission-accounts id])
+        {:keys [highest-permission-role]} (rf/sub [:community/token-gated-overview id])
+        highest-permission-role-text
```
`highest-permission-role-*` is a very clear name. Perhaps shortening it to just `highest-role-*` gives the same clarity.

yqrashawn replied:
changed to highest-role-text

ilmotta commented on src/status_im/contexts/communities/events.cljs:
```diff
@@ -64,33 +64,56 @@
-               [:dispatch [:communities/check-permissions-to-join-community id]])
-             (when (some has-channel-perm? token-permissions)
-               [:dispatch [:communities/check-all-community-channels-permissions id]])]}))))
+(defn <-revealed-accounts-rpc
```
To follow our current standards, this namespace already contains logic that should be moved to the data store namespaces, particularly the `*-rpc` functions. It would be nice if you could write the new functions already in the data store directory, such as `<-revealed-accounts-rpc`.

yqrashawn replied:
moved all these <- fn to legacy.status-im.data-store.communities

ilmotta commented on src/status_im/contexts/communities/events_test.cljs:
```diff
@@ -193,3 +194,88 @@
+        (is (match?
+             [[:dispatch [:communities/initialize-permission-addresses community-id]]
+              [:dispatch [:chat.ui/spectate-community community-id]]
+              nil
```
It's a bit hard to reason what these nils represent without having to check the implementation and the order of `when` usages. We could have a margin comment on each one, or maybe better just remove nils and only assert what actually gets dispatched.

yqrashawn replied:
removed with filter some?

ilmotta commented on src/status_im/subs/communities.cljs:
```diff
@@ -297,32 +297,62 @@
+                            constants/community-token-permission-become-token-owner}
+         highest-permission-role
+         (when can-request-access?
+           (->> token-permissions-check
```
This subscription is turning out to be somewhat complex to understand. I don't know if it's worth refactoring it too much because we can actually consider having a single endpoint in status-go to return all of this, but since we don't have that yet, consider refactoring it into small private functions in this sub namespace, this would go a long of helping understand the code.

yqrashawn replied:
The permissions and permissions-check in db community are in weird shapes (maybe there's a reason?) and do need some refactors on both mobile and go side

> having a single endpoint in status-go to return all of this

the approach using `wakuext_fetchCommunity` to re-fetch community should be removed and rely on status-go signal status-mobile about community updates
I think permission updates should be handled that way as well (already partially handled? since the video shows mobile react and perm change automatically)

`:communities/check-permissions-to-join-community` (to get token-permissions-check) do many requests to blockchain nodes on every call
which might be a problem if merge it into the same endpoint

maybe balance change -> go signals changes to mobile is a better approach
since after the include all future address feature new wallet account will also trigger changes

ilmotta replied:
Good points. I don't know either if the signals are properly handled by us, but looks like it as you said. Something to be checked by us.

I agree with you @yqrashawn, the pull approach to get balances is usually not a good idea. We could cache on the backend, but at the end of the line to join a community, some check(s) against chain(s) must be performed anyway.

The endpoint to provide what I call "permissioned balances" (relevant tokens per Figma) I'm working is also implemented by pulling data from the chains. I haven't looked at a signals approach, but now that you say, I think we should, although I think the machinery is not fully ready in status-go. For instance, the permissions checker in status-go does sync calls to the chains on every check.

ilmotta commented on src/status_im/contexts/communities/events.cljs:
```diff
@@ -405,3 +436,54 @@
+
+(rf/reg-event-fx :communities/get-revealed-accounts get-revealed-accounts)
+
+(schema/=> get-revealed-accounts
```
👏🏼 👏🏼

ilmotta reviewed (commented):
@yqrashawn, since the PR is still in progress, I did a quick review pass. Let me know once you have fully finished it and I can review again. Thank you!

ilmotta commented on src/status_im/contexts/communities/events.cljs:
```diff
@@ -64,33 +64,56 @@
       (update :token-permissions seq)
       (update :categories <-categories-rpc)
+      (assoc :membership-permissions?
+             (some #(= (:type %) constants/community-token-permission-become-member)
```
This solution here is probably okay for your PR, but as we know @ajayesivan is also working in a PR that is trying to fix all places that need to consume the sequence of token permissions. Perhaps after you merge this one, or if Ajay merges his one first we should try to use a unified predicate to tell if there are membership permissions, just so we avoid bugs because we might forget to update all places with duplicated logic.

ilmotta reviewed (approved):
Nice work @yqrashawn. There's some room for some refactorings as I suggested, but nothing that can't be done in future PRs related to account selection.

yqrashawn committed:
refactor: based on PR feedback

- move <-rpc to data store
- `role->translation-key`
- highest-permission-role-text -> highest-role-text
- filter out nils for handle community test

Signed-off-by: yqrashawn <namy.19@gmail.com>
